### PR TITLE
Support remote storage (fsspec URLs) for FSDP checkpoints

### LINKED ISSUE #21779: Support remote/cloud storage (fsspec) for FSDP checkpointing
### Description & Motivation

Currently, FSDP (Fully Sharded Data Parallel) checkpointing in Lightning (both Fabric and PyTorch strategies) assumes that the checkpoint path is a local filesystem path. This leads to several issues when attempting to save or load checkpoints directly to/from remote object storage (e.g., `gs://bucket/path`, `s3://bucket/path`):

1. **Path Corruption**: The code uses `pathlib.Path(path)` on the broadcasted path. For remote URLs, this collapses the double slash (e.g., `gs://` becomes `gs:/`), which corrupts the URL and makes it invalid for cloud storage providers.
2. **Hardcoded Local Readers/Writers**: For PyTorch versions < 2.3, FSDP checkpointing explicitly instantiates `FileSystemWriter` and `FileSystemReader`, which only support local paths. PyTorch's Distributed Checkpoint (DCP) has support for remote filesystems via `FsspecWriter` and `FsspecReader`, but these are not utilized.
3. **Incompatible Local File Operations**: Operations like `path.is_dir()`, `path.mkdir()`, `path.unlink()`, and `shutil.rmtree()` are used directly. These do not work on remote object storage.
4. **Unsupported `mmap` on Remote Storage**: When loading a full checkpoint, the code attempts to use `mmap=True` or `_lazy_load`, which requires local file access and fails over object storage.

We need to update the FSDP strategies to support remote paths seamlessly using `fsspec`.

## Proposed Solution

We propose to introduce cloud-aware path handling and utility functions, and update the FSDP strategies to use them.

### 1. Cloud-Aware Path Utilities (`lightning/fabric/utilities/cloud_io.py`)
Introduce the following helper functions to abstract filesystem operations:
*   `_resolve_path(path)`: Returns a `pathlib.Path` for local paths, but keeps remote `fsspec` URLs as `str` to prevent corruption.
*   `_checkpoint_join(path, name)`: Safely joins a filename/directory name to a path (using `/` operator for `Path`, or string concatenation for remote URLs).
*   `_is_checkpoint_dir(path)`: Checks if a path is a directory, supporting `fsspec` paths via `get_filesystem(path).isdir`.
*   `_prepare_directory_checkpoint(path)`: Prepares a directory for a sharded checkpoint. If a file exists at the path, it removes it. It creates the directory if it's a local path or a non-object remote path (directory creation is a no-op on object storage).
*   `_remove_checkpoint(path)`: Recursively removes a checkpoint file or directory, supporting `fsspec` paths via `fs.rm(..., recursive=True)`.

### 2. Update FSDP Strategies (Fabric & PyTorch)
*   **Path Resolution**: Use `_resolve_path` on the broadcasted checkpoint path instead of wrapping it in `Path` directly.
*   **Directory/File Checks & Preparation**: Replace `path.is_dir()`, `path.mkdir()`, and `shutil.rmtree()` with `_is_checkpoint_dir()`, `_prepare_directory_checkpoint()`, and `_remove_checkpoint()`.
*   **DCP Reader/Writer Integration**:
    *   Implement `_get_distributed_checkpoint_writer(path)` and `_get_distributed_checkpoint_reader(path)`.
    *   If the path is a remote URL, return `FsspecWriter`/`FsspecReader` (from `torch.distributed.checkpoint._fsspec_filesystem`).
    *   If the path is local, fall back to `FileSystemWriter`/`FileSystemReader`.
    *   For PyTorch >= 2.3, we can let `torch.distributed.checkpoint.save/load` automatically infer the writer/reader from the `checkpoint_id` (which now supports `fsspec` paths). For older versions, use the helper-acquired writer/reader.
*   **Metadata Saving/Loading**: Use `_atomic_save` and `_load` combined with `_checkpoint_join` for the metadata file (`meta.pt`).

### 3. Update Model Parallel Strategy (`lightning/fabric/strategies/model_parallel.py`)
*   In `_load_raw_module_state_from_path`, if the path is remote (detected via `_is_local_file_protocol`), bypass `mmap` / `_lazy_load` and load the state dict using `_load(path, map_location="cpu")`.

### Pitch

_No response_

### Alternatives

_No response_

### Additional context

_No response_

cc @lantiga

### PR DESCRIPTION
**Pull request by zhixiangli**
## What does this PR do?

Fix https://github.com/Lightning-AI/pytorch-lightning/issues/21779

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

**Comment by zhixiangli** [comment]
Hi @tchaton @justusschock @ethanwharris

I'm eager to get this PR in shape, but I'm currently having trouble accessing the CI failure logs. 

When I follow the check links, I hit a login page. I attempted to register for a Lightning AI account to view them, but the phone verification step is consistently rejecting my Singapore phone number with this error:
> **Invalid number (support code: 09332104)**

I have already reached out to `support@lightning.ai` with my phone details. Could someone kindly let me know if registering for an account is the correct path for external contributors to view these logs, or if special access needs to be granted in another way? 

If anyone on the team is able to look into the support ticket or safely expedite the verification process, I would immensely appreciate it.

Thank you so much for your time and help!

File: src/lightning/fabric/utilities/cloud_io.py [hunk 1]
```diff
@@ -176,3 +178,98 @@ def _is_dir(fs: AbstractFileSystem, path: Union[str, Path], strict: bool = False
...
+
+        return FileSystemReader(path=path)
+    FsspecReader = _import_fsspec_dcp_filesystem("FsspecReader")
+    return FsspecReader(path=str(path))
+
+
+def _import_fsspec_dcp_filesystem(name: str) -> Any:
+    """Import ``FsspecReader``/``FsspecWriter`` from torch's private DCP fsspec module.
+
+    These live in a private module that not every PyTorch build ships, so raise an actionable error
```
Comment (ankitaluthra1) [inline src/lightning/fabric/utilities/cloud_io.py:267]: This would create a maintenance risk; PyTorch could change or remove these classes in future releases without warning.
I have asked to make FsspecReader/FsspecWriter public in pytorch
https://github.com/pytorch/pytorch/pull/188698
consider merging these changes afterwards

File: src/lightning/fabric/strategies/model_parallel.py [hunk 2]
```diff
@@ -529,15 +530,18 @@ def _has_dtensor_modules(module: object) -> TypeGuard[Module]:
...
             "Failed to load checkpoint directly into the model. The given path must be a single file containing the"
             f" full state dict: {path}"
         )
-    # Use `lazy_load`/`mmap` instead to avoid storing a copy of the full checkpoint per rank
-    state_dict = torch.load(path, mmap=True, map_location="cpu") if _TORCH_GREATER_EQUAL_2_3 else _lazy_load(path)
+    if _is_local_file_protocol(str(path)):
+        # Use `lazy_load`/`mmap` instead to avoid storing a copy of the full checkpoint per rank
+        state_dict = torch.load(path, mmap=True, map_location="cpu") if _TORCH_GREATER_EQUAL_2_3 else _lazy_load(path)
+    else:
+        state_dict = _load(path, map_location="cpu")
```
Comment (ankitaluthra1) [inline src/lightning/fabric/strategies/model_parallel.py:544]: IIUC , lazy loading was done to avoid loading the full checkpoint into memory across all ranks. since we are not using mmap for remote checkpoint, this would mean reading remote checkpoints will fully load them into memory, significantly increasing peak memory usage.
Is it possible to use [mmap caching](https://github.com/fsspec/filesystem_spec/blob/82d24f76f39fbc45a6a18af5d98d88a785bc43f9/fsspec/caching.py#L128) technique here to aid lazy loading? If not, this needs to be called out clearly.

File: src/lightning/fabric/utilities/cloud_io.py [hunk 3]
```diff
@@ -176,3 +178,98 @@ def _is_dir(fs: AbstractFileSystem, path: Union[str, Path], strict: bool = False
 
 def _is_local_file_protocol(path: _PATH) -> bool:
     return fsspec.utils.get_protocol(str(path)) == "file"
+
+
+def _resolve_path(path: _PATH) -> Union[str, Path]:
```
Comment (ankitaluthra1) [inline src/lightning/fabric/utilities/cloud_io.py:183]: Since other parts of the codebase expect a strict Path object before these changes, please make sure all callers remain compatible for remote as well where str might be returned

File: src/lightning/fabric/utilities/cloud_io.py (hunk 1 above)
Comment (zhixiangli) [reply src/lightning/fabric/utilities/cloud_io.py:267]: This project pins the PyTorch version to torch >=2.1.0, <2.10.0 to prevent breaking changes, and future regressions will be caught by tests. While the new changes from https://github.com/pytorch/pytorch/pull/188698 will be very helpful down the road, they won't be picked up right now due to the version pin. Therefore, this PR does not depend on the pytorch PR.

File: src/lightning/fabric/utilities/cloud_io.py (hunk 3 above)
Comment (zhixiangli) [reply src/lightning/fabric/utilities/cloud_io.py:183]: All the operations that previously assumed a Path were converted to str/Path-agnostic helpers.

File: src/lightning/fabric/utilities/cloud_io.py [hunk 4]
```diff
@@ -176,3 +178,98 @@ def _is_dir(fs: AbstractFileSystem, path: Union[str, Path], strict: bool = False
...
+    if _is_local_file_protocol(str(path)):
+        return Path(path)
+    return str(path)
+
+
+def _checkpoint_join(path: Union[str, Path], name: str) -> Union[str, Path]:
+    """Join ``name`` onto a checkpoint ``path`` without corrupting remote URLs."""
+    if isinstance(path, Path):
+        return path / name
+    return str(path).rstrip("/") + "/" + name
```
Comment (deependujha) [inline src/lightning/fabric/utilities/cloud_io.py:203]: shouldn't we use `os.path.join` so it also doesn't fail on windows

File: src/lightning/fabric/strategies/model_parallel.py (hunk 2 above)
Comment (zhixiangli) [reply src/lightning/fabric/strategies/model_parallel.py:544]: That’s a good catch! I think a feasible solution is to download the data, save it as a temporary file, and pass the file descriptor to PyTorch's mmap.

However, I’d like to keep the scope of this PR limited to fsspec integration, if you don't mind. As long as it doesn't break existing features, I believe we should keep this PR as focused as possible while it is in the early stages. We can prioritize improving this new fsspec path in a follow-up issue and PR once we have the benchmarks.

File: src/lightning/fabric/utilities/cloud_io.py (hunk 4 above)
Comment (zhixiangli) [reply src/lightning/fabric/utilities/cloud_io.py:203]: If use os.path.join, it would break remote storage support on Windows.

Remote URLs (e.g., s3://..., gs://...):
- On Windows, os.path.join("gs://bucket", "metadata") uses the Windows path separator (), resulting in gs://bucket\metadata, which is an invalid URL.
- Remote URLs must always use / as the separator.

Comment (deependujha) [reply src/lightning/fabric/utilities/cloud_io.py:203]: my bad 😅 thanks :)

File: src/lightning/fabric/strategies/fsdp.py [hunk 5]
```diff
@@ -600,7 +608,7 @@ def load_checkpoint(
             return metadata
 
         if _is_full_checkpoint(path):
-            checkpoint = _lazy_load(path)
+            checkpoint = _lazy_load(path) if _is_local_file_protocol(str(path)) else _load(path, weights_only=False)
```
[the remark this answers is not in the document]
Comment (zhixiangli) [reply src/lightning/fabric/strategies/fsdp.py:626]: Done

File: src/lightning/pytorch/strategies/fsdp.py [hunk 6]
```diff
@@ -626,11 +631,11 @@ def load_checkpoint(self, checkpoint_path: _PATH, weights_only: Optional[bool] =
...
 
             # Load metadata (anything not a module or optimizer)
-            metadata = torch.load(path / _METADATA_FILENAME, weights_only=weights_only)
+            metadata = _load(_checkpoint_join(path, _METADATA_FILENAME), weights_only=weights_only)
             return metadata
 
         if _is_full_checkpoint(path):
-            checkpoint = _lazy_load(path)
+            checkpoint = _lazy_load(path) if _is_local_file_protocol(str(path)) else _load(path, weights_only=False)
             _load_raw_module_state(
```
[the remark this answers is not in the document]
Comment (zhixiangli) [reply src/lightning/pytorch/strategies/fsdp.py:653]: Done

File: src/lightning/fabric/utilities/cloud_io.py [hunk 7]
```diff
@@ -176,3 +178,98 @@ def _is_dir(fs: AbstractFileSystem, path: Union[str, Path], strict: bool = False
...
+def _resolve_path(path: _PATH) -> Union[str, Path]:
+    """Return a ``Path`` for local file paths and a plain ``str`` for remote fsspec URLs.
+
+    ``Path()`` collapses the double slash in a URL (e.g. ``gs://bucket`` -> ``gs:/bucket``),
+    corrupting it, so remote URLs must be kept as strings.
+
+    """
+    if _is_local_file_protocol(str(path)):
+        return Path(path)
+    return str(path)
```
[the remark this answers is not in the document]
Comment (zhixiangli) [reply src/lightning/fabric/utilities/cloud_io.py:193]: Done

File: src/lightning/fabric/utilities/cloud_io.py [hunk 8]
```diff
@@ -176,3 +178,99 @@ def _is_dir(fs: AbstractFileSystem, path: Union[str, Path], strict: bool = False
...
+        _, urlpath = url_to_fs(str(path))
+        return Path(urlpath)
+    return str(path)
+
+
+def _checkpoint_join(path: Union[str, Path], name: str) -> Union[str, Path]:
+    """Join ``name`` onto a checkpoint ``path`` without corrupting remote URLs."""
+    if isinstance(path, Path):
+        return path / name
+    return str(path).rstrip("/") + "/" + name
```
Comment (deependujha) [inline src/lightning/fabric/utilities/cloud_io.py:203]: this is fine with context, but just looking at this function might make one wonder (atleast i got confused) why not replace it with `os.path.join`, so a comment would be nice

File: src/lightning/fabric/utilities/cloud_io.py (hunk 1 above)
Comment (deependujha) [reply src/lightning/fabric/utilities/cloud_io.py:267]: hi @zhixiangli, we keep upgrading pytorch version, and i doubt if there's a relevant regression test. A minimal test that just tries importing it on `torch>=2.6` should catch it early when we upgrade the pinned version

Comment (deependujha) [reply src/lightning/fabric/utilities/cloud_io.py:267]: sorry, verified it exists in `test_distributed_checkpoint_reader_writer_selection` :)

**Review by deependujha — APPROVED** [review]

**Comment by deependujha** [comment]
We need to update `load_sharded_optimizer_state_dict` for `torch>=2.3`.

```
/home/zeus/miniconda3/envs/cloudspace/lib/python3.12/site-packages/torch/distributed/checkpoint/optimizer.py:348: FutureWarning: `load_state_dict` is deprecated and will be removed in future versions. Please use `load` instead.
```

I'll handle it in a separate PR. verified with PL on T4 4gpu with saving sharded checkpoints on AWS S3. ✅

<details>

<summary>Code used</summary>

```python
from typing import override
import torch
import warnings
import torch.nn as nn
from torch.utils.data import TensorDataset, DataLoader
import lightning as L
from lightning.pytorch.strategies import FSDPStrategy

L.seed_everything(42)

warnings.filterwarnings(
    "ignore",
    message=".*Please use DTensor instead.*",
    category=FutureWarning,
)
warnings.filterwarnings(
    "ignore",
    message=".*_get_pg_default_device.*",
)

class MySuperCoolDataModule(L.LightningDataModule):
    def __init__(
        self,
        batch_size=32,
        dataset_size=64,
        input_dim=8,
    ):
        super().__init__()
        self.save_hyperparameters()

        self.batch_size = batch_size
        self.x = torch.randn(dataset_size, input_dim)
        self.y = torch.randn(dataset_size)

    def setup(self, stage=None):
        self.dataset = TensorDataset(self.x, self.y)

    def train_dataloader(self):
        return DataLoader(
            self.dataset,
            batch_size=self.batch_size,
            shuffle=True,
            num_workers=2,
        )


class MySuperCoolModel(L.LightningModule):
    def __init__(self, input_dim=8):
        super().__init__()
        self.save_hyperparameters()

        self.layers = nn.Sequential(
            nn.Linear(input_dim, 256),
            nn.ReLU(),
            nn.Linear(256, 2048),
            nn.ReLU(),
            nn.Linear(2048, 2048),
            nn.ReLU(),
            nn.Linear(2048, 2048),
            nn.ReLU(),
            nn.Linear(2048, 256),
            nn.ReLU(),
            nn.Linear(256, 32),
            nn.ReLU(),
            nn.Linear(32, 1),
        )

    def forward(self, x):
        return self.layers(x)

    def training_step(self, batch, batch_idx):
        x, y = batch

        y_hat = self(x).squeeze(-1)
        loss = nn.functional.mse_loss(y_hat, y)

        self.log("train_loss", loss)

        return loss

    @override
    def configure_optimizers(self):
        return torch.optim.Adam(
            self.parameters(),
            lr=1e-3,
        )


def main():
    model = MySuperCoolModel()

    datamodule = MySuperCoolDataModule(
        batch_size=32,
        dataset_size=64,
        input_dim=8,
    )

    policy = {nn.Linear,}

    trainer = L.Trainer(
        default_root_dir="s3://lightning-checkpoints/simple-checkpoint-02/",
        accelerator="auto",
        devices="auto",
        strategy=FSDPStrategy(auto_wrap_policy=policy, state_dict_type="sharded"),
        precision="32-true",
        max_epochs=2,
        log_every_n_steps=1,
    )

    trainer.fit(model, datamodule)

    trainer.print("-" * 80)
    trainer.print("Training done. ✅")
    # trainer.print(torch.cuda.memory_summary())


if __name__ == "__main__":
    main()
```
</details>

**Comment by zhixiangli** [comment]
Hi @deependujha , thank you for the approval! Just checking in to see if there are any remaining steps needed on my end, or if this is ready to be merged. Thanks again for your time and review!

cc: @tchaton, @justusschock, @ethanwharris

**Review by tchaton — APPROVED** [review]
Nice ;)
